internal/logging/file.go

```diff
@@ -0,0 +1,220 @@
+package logging
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/ActiveState/cli/internal/condition"
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/rollbar/rollbar-go"
+)
+
+var defaultMaxEntries = 1000
+
+type entry struct {
+	ctx     *MessageContext
+	message string
+	args    []interface{}
+}
+
+type fileHandler struct {
+	formatter Formatter
+	file      *os.File
+	mu        sync.Mutex
+	verbose   safeBool
+	wg        *sync.WaitGroup
+	queue     chan entry
+	quit      chan struct{}
+}
+
+func newFileHandler() *fileHandler {
+	handler := fileHandler{
+		DefaultFormatter,
+		nil,
+		sync.Mutex{},
+		safeBool{},
+		&sync.WaitGroup{},
+		make(chan entry, defaultMaxEntries),
+		make(chan struct{}),
+	}
+	handler.wg.Add(1)
+	go func() {
+		defer handler.wg.Done()
+		handler.start()
+	}()
+	return &handler
+}
+
+func (l *fileHandler) start() {
+	for {
+		select {
+		case entry := <-l.queue:
+			l.emit(entry.ctx, entry.message, entry.args...)
+		case <-l.quit:
+			close(l.queue)
+			for entry := range l.queue {
+				l.emit(entry.ctx, entry.message, entry.args...)
+			}
+			return
+		}
+	}
+}
+
+func (l *fileHandler) SetFormatter(f Formatter) {
+	l.formatter = f
+}
+
+func (l *fileHandler) SetVerbose(v bool) {
+	l.verbose.setValue(v)
+}
+
+func (l *fileHandler) Output() io.Writer {
+	return l.file
+}
+
+func (l *fileHandler) Emit(ctx *MessageContext, message string, args ...interface{}) error {
+	var a []interface{}
+	a = append(a, args)
+	e := entry{
+		ctx:     ctx,
+		message: message,
+		args:    a,
+	}
+	select {
+	case <-l.quit:
+		return nil
+	default:
+		l.queue <- e
+	}
+	return nil
+}
+
+func (l *fileHandler) emit(ctx *MessageContext, message string, args ...interface{}) {
+	defer handlePanics(recover())
+	// In this function we close and open the file handle to the log file. In
+	// order to ensure this is safe to be called across threads, we just
+	// synchronize the entire function
+	l.mu.Lock()
+	defer l.mu.Unlock()
```

Comment:
This doesn't seem necessary anymore since we're processing them on a queue. @daved can you confirm my assumption?

Reply:
I think it's not needed for the current design, but I take the file sync to be distinct from the log emission synchronization. Also, I'd like to see a test to be sure if race detection or output verification (less likely) show that the mutex is needed or not. If so, the mutex hold-time should be decreased (likely by wrapping the file in a struct to wrap the Write/Read methods so that the mutex is held as little as possible). Maybe it should be left for now and a follow-up story filed.

Reply:
Added a unit test and ran it with the `-race` flag successfully.